articles/143_legacy_interface_definition.md

```diff
@@ -38,22 +38,30 @@ Each field is described by a *type* and a *name*.
 
 A single data structure is called *message*.
 Each message has a *name*.
-Together with the name of the *package* a message can be uniquely identified.
 
 ### Services
 
 For request / reply style communication the two exchanged data structures are related.
 These pairs of data structures are called *services*.
-A service is identified by its *name* and the *package* it is in.
 Each service describes two messages, one for the request data structure, one for the reply data structure.
 
 ### Actions
 
 For longer running request / reply style communication with feedback about the progress the exchanged data structures are related.
 These triplets of data structures are called *actions*.
-An action is identified by its *name* and the *package* it is in.
 Each action describes three messages, one for the goal data structure, one for the result data structure, and one for the feedback data structure.
 
+### Identifying data structures
+
+Every data structure can be uniquely identified with three pieces of information:
+
+1. **Package** - the name of the package containing the data structure definition.
+2. **Subfolders** - the list of subfolders within the package where the data structure defintion can be found.
```

Comment:
I'm not keen on needing to keep the internal file structure of my package constant forever just to avoid breaking a name.

Reply:
The information is not internal. The subfolder is part of the namespace of the data structure.

Reply:
I think this can be reworded and perhaps subfolder is not the best word to use. What about "interface namespace"?

Reply:
Maybe just `subnamespaces`?

Reply:
Updated to use "subnamespaces" 11564de

Reply:
You will need to specify how a subnamespace is defined now, which comes back to the use of folders, i.e. package structure, again.

Perhaps we need a way to specify the message definition root in a package and all folders below that are treated as name spaces?

Either that, or we state outright that best practice is that messages should live in a subfolder of the package called `msg`, etc. so that people are likely to get it right from the start and not need to change it. I wouldn't be opposed to that.

Reply:
I did some experimenting. IIUC, as far as typesupport is concerned, it doesn't matter where you place message definitions in a package. For example, `msg/Foo.msg` and `some/other/subfolder/Foo.msg` will generate the same files under the `msg` namespace. Likewise for services and actions (I guess the namespace is determined by the file suffix, `.msg`, `.srv`, and `.action`?). @dirk-thomas Is my understanding correct?

If it has any consequence, I did note that the original definition files (not the `.idl` files) appear to be installed in a directory named after the parent directory. In my example above, I see `subfolder/Foo.msg` in the packages `share` directory.